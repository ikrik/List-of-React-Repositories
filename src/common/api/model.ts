export interface IPageInfo {
  endCursor?: string
  hasNextPage?: boolean
  hasPreviousPage?: boolean
  startCursor?: string
}

export interface ISearch<T> {
  edges: T[]
  pageInfo?: IPageInfo
}
