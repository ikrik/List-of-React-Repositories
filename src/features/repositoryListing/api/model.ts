import { ISearch } from '../../../common/api/model'

export interface IRepositoryInfo {
  forkCount: number
  id: string
  name: string
  stargazerCount: number
  url: string
}

export interface IRepositoryListingItem {
  cursor: string
  node: IRepositoryInfo
}

export interface IRepositoryListingSearch {
  search: ISearch<IRepositoryListingItem>
}
