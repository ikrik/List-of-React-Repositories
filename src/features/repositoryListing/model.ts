import { TableColumnDescriptor } from '../../common/models'

export interface RepositoryListingTableView {
  id: string
  forkCount: number
  name: string
  stargazerCount: number
  url?: string
}

export type RepositoryListingDescriptor = {
  [P in keyof RepositoryListingTableView]: TableColumnDescriptor<RepositoryListingTableView>
}
