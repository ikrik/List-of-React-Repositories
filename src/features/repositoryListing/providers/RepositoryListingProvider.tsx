import * as React from 'react'
import { RepositoryListingTableView } from '../model'
import useRepositoryListingData, { DEFAULT_REPO_LISTING_DATA_INTERFACE } from './hooks/useRepositoryListingData'

export interface IRepositoryListing {
  isRepositoryListLoading: boolean
  repositoryListing: RepositoryListingTableView[]
  listSize: number
  totalPages: number
  pageNumber: number
  disableNextPage: boolean
  disablePreviousPage: boolean
  onNextPageNavigation: () => void
  onPreviousPageNavigation: () => void
}

export const DEFAULT_REPOSITORY_LISTING_STORE: IRepositoryListing = {
  ...DEFAULT_REPO_LISTING_DATA_INTERFACE,
  listSize: 20,
  repositoryListing: DEFAULT_REPO_LISTING_DATA_INTERFACE.mappedRepositoryListing,
  onNextPageNavigation: DEFAULT_REPO_LISTING_DATA_INTERFACE.onNextPage,
  onPreviousPageNavigation: DEFAULT_REPO_LISTING_DATA_INTERFACE.onPreviousPage
}

export const RepositoryListingContext = React.createContext(DEFAULT_REPOSITORY_LISTING_STORE)

const RepositoryListingProvider: React.FunctionComponent = (props) => {
  const [listSize, setListSize] = React.useState<number>(DEFAULT_REPOSITORY_LISTING_STORE.listSize)
  const repoListiDataStore = useRepositoryListingData(listSize)

  return (
    <RepositoryListingContext.Provider
      value={{
        isRepositoryListLoading: repoListiDataStore.isRepositoryListLoading,
        repositoryListing: repoListiDataStore.mappedRepositoryListing,
        listSize,
        totalPages: repoListiDataStore.totalPages,
        pageNumber: repoListiDataStore.pageNumber,
        disableNextPage: repoListiDataStore.disableNextPage,
        disablePreviousPage: repoListiDataStore.disablePreviousPage,
        onNextPageNavigation: repoListiDataStore.onNextPage,
        onPreviousPageNavigation: repoListiDataStore.onPreviousPage
      }}
    >
      {props.children}
    </RepositoryListingContext.Provider>
  )
}

export default RepositoryListingProvider
