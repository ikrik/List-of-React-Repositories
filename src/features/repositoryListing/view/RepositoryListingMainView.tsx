import * as React from 'react'
import styled from 'styled-components'
import { RepositoryListingContext } from '../providers/RepositoryListingProvider'
import RepositoryListingPagination from './RepositoryListingPagination'
import RepositoryListingTable from './RepositoryListingTable'

const RepositoryListingMainWrapper = styled.div`
  width: 100%;
  max-width: 800px;
  align-self: center;
`

const RepositoryListingMainView: React.FunctionComponent = () => {
  const {
    isRepositoryListLoading,
    totalPages,
    pageNumber,
    disableNextPage,
    disablePreviousPage,
    onPreviousPageNavigation,
    onNextPageNavigation
  } = React.useContext(RepositoryListingContext)
  return (
    <RepositoryListingMainWrapper>
      <RepositoryListingTable />
      <RepositoryListingPagination
        totalPages={totalPages}
        pageNumber={pageNumber}
        disableNextPage={disableNextPage}
        disablePreviousPage={disablePreviousPage}
        nextPage={onNextPageNavigation}
        previousPage={onPreviousPageNavigation}
      />
    </RepositoryListingMainWrapper>
  )
}

export default RepositoryListingMainView
