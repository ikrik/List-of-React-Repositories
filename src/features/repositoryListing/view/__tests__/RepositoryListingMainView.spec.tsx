import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import { RepositoryListingTableView } from '../../model'

import { DEFAULT_REPOSITORY_LISTING_STORE, RepositoryListingContext } from '../../providers/RepositoryListingProvider'
import RepositoryListingMainView from '../RepositoryListingMainView'

describe('Manual Inputs > Trade Op main view component tests', () => {
  afterEach(cleanup)

  it('renders empty state when payments data is not yet fetched', () => {
    const wrapper = render(
      <RepositoryListingContext.Provider
        value={{
          ...DEFAULT_REPOSITORY_LISTING_STORE,
          repositoryListing: [],
          isRepositoryListLoading: false
        }}
      >
        <RepositoryListingMainView />
      </RepositoryListingContext.Provider>
    )

    expect(wrapper.container.firstChild).toMatchSnapshot()
  })

  it('renders screen loader when the data is loading', () => {
    const wrapper = render(
      <RepositoryListingContext.Provider
        value={{
          ...DEFAULT_REPOSITORY_LISTING_STORE,
          isRepositoryListLoading: true
        }}
      >
        <RepositoryListingMainView />
      </RepositoryListingContext.Provider>
    )

    expect(wrapper.container.firstChild).toMatchSnapshot()
  })

  it('renders payments inputs table when the data is loaded', () => {
    const mockMappedRepositoryListing: RepositoryListingTableView[] = [
      {
        id: 'ydhsnmmXjash=',
        forkCount: 988327,
        name: 'React-native',
        stargazerCount: 93288,
        url: 'https://github.com/React-native'
      },
      {
        id: 'ydeweXjash=',
        forkCount: 32756,
        name: 'babel-react',
        stargazerCount: 156788,
        url: 'https://github.com/babel-react'
      },
      {
        id: 'ydhggmmXjash=',
        forkCount: 34998,
        name: 'next.js',
        stargazerCount: 288,
        url: 'https://github.com/next'
      }
    ]
    const wrapper = render(
      <RepositoryListingContext.Provider
        value={{
          ...DEFAULT_REPOSITORY_LISTING_STORE,
          repositoryListing: mockMappedRepositoryListing
        }}
      >
        <RepositoryListingMainView />
      </RepositoryListingContext.Provider>
    )

    expect(wrapper.container.firstChild).toMatchSnapshot()
  })
})
