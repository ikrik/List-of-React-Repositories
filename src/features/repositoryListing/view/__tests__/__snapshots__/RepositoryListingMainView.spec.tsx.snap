// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manual Inputs > Trade Op main view component tests renders empty state when payments data is not yet fetched 1`] = `
<div
  class="sc-jSgupP kpIYLA"
>
  <table
    class="Table-oyilxr-0 jQhyhB table"
  >
    <caption
      class="Caption-sc-1miu1rz-0 eQOCZM"
    >
      List of Github React Repositories
    </caption>
    <thead
      class=""
    >
      <tr
        class="table-active"
      >
        <th
          class=""
        >
          
          Name
        </th>
        <th
          class=""
        >
          <svg
            height="24px"
            style="vertical-align: sub;"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g
              data-name="Layer 2"
            >
              <g
                data-name="star"
              >
                <rect
                  height="24"
                  opacity="0"
                  transform="rotate(90 12 12)"
                  width="24"
                />
                <path
                  d="M17.56 21a1 1 0 0 1-.46-.11L12 18.22l-5.1 2.67a1 1 0 0 1-1.45-1.06l1-5.63-4.12-4a1 1 0 0 1-.25-1 1 1 0 0 1 .81-.68l5.7-.83 2.51-5.13a1 1 0 0 1 1.8 0l2.54 5.12 5.7.83a1 1 0 0 1 .81.68 1 1 0 0 1-.25 1l-4.12 4 1 5.63a1 1 0 0 1-.4 1 1 1 0 0 1-.62.18z"
                  fill="#fbcc32"
                />
              </g>
            </g>
          </svg>
          Stars
        </th>
        <th
          class=""
        >
          <svg
            height="24px"
            style="vertical-align: sub;"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <path
              d="M17,6.06a3,3,0,0,0-1.15,5.77A2,2,0,0,1,14,13.06H10a3.91,3.91,0,0,0-2,.56V7.88a3,3,0,1,0-2,0v8.36a3,3,0,1,0,2.16.05A2,2,0,0,1,10,15.06h4a4,4,0,0,0,3.91-3.16A3,3,0,0,0,17,6.06Zm-10-2a1,1,0,1,1-1,1A1,1,0,0,1,7,4.06Zm0,16a1,1,0,1,1,1-1A1,1,0,0,1,7,20.06Zm10-10a1,1,0,1,1,1-1A1,1,0,0,1,17,10.06Z"
            />
          </svg>
          Forks
        </th>
      </tr>
    </thead>
    <tbody
      class="Tbody-sc-14l4jy7-0"
    />
  </table>
</div>
`;

exports[`Manual Inputs > Trade Op main view component tests renders payments inputs table when the data is loaded 1`] = `
<div
  class="sc-jSgupP kpIYLA"
>
  <table
    class="Table-oyilxr-0 jQhyhB table"
  >
    <caption
      class="Caption-sc-1miu1rz-0 eQOCZM"
    >
      List of Github React Repositories
    </caption>
    <thead
      class=""
    >
      <tr
        class="table-active"
      >
        <th
          class=""
        >
          
          Name
        </th>
        <th
          class=""
        >
          <svg
            height="24px"
            style="vertical-align: sub;"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g
              data-name="Layer 2"
            >
              <g
                data-name="star"
              >
                <rect
                  height="24"
                  opacity="0"
                  transform="rotate(90 12 12)"
                  width="24"
                />
                <path
                  d="M17.56 21a1 1 0 0 1-.46-.11L12 18.22l-5.1 2.67a1 1 0 0 1-1.45-1.06l1-5.63-4.12-4a1 1 0 0 1-.25-1 1 1 0 0 1 .81-.68l5.7-.83 2.51-5.13a1 1 0 0 1 1.8 0l2.54 5.12 5.7.83a1 1 0 0 1 .81.68 1 1 0 0 1-.25 1l-4.12 4 1 5.63a1 1 0 0 1-.4 1 1 1 0 0 1-.62.18z"
                  fill="#fbcc32"
                />
              </g>
            </g>
          </svg>
          Stars
        </th>
        <th
          class=""
        >
          <svg
            height="24px"
            style="vertical-align: sub;"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <path
              d="M17,6.06a3,3,0,0,0-1.15,5.77A2,2,0,0,1,14,13.06H10a3.91,3.91,0,0,0-2,.56V7.88a3,3,0,1,0-2,0v8.36a3,3,0,1,0,2.16.05A2,2,0,0,1,10,15.06h4a4,4,0,0,0,3.91-3.16A3,3,0,0,0,17,6.06Zm-10-2a1,1,0,1,1-1,1A1,1,0,0,1,7,4.06Zm0,16a1,1,0,1,1,1-1A1,1,0,0,1,7,20.06Zm10-10a1,1,0,1,1,1-1A1,1,0,0,1,17,10.06Z"
            />
          </svg>
          Forks
        </th>
      </tr>
    </thead>
    <tbody
      class="Tbody-sc-14l4jy7-0"
    >
      <tr
        class=""
      >
        <td
          align="Name"
          class=""
        >
          <a
            class="A-sc-70cat-0 ljzaKq"
            href="https://github.com/React-native"
            target="_blank"
          >
            React-native
          </a>
        </td>
        <td
          align="Stars"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            93,288
          </span>
        </td>
        <td
          align="Forks"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            988,327
          </span>
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          align="Name"
          class=""
        >
          <a
            class="A-sc-70cat-0 ljzaKq"
            href="https://github.com/babel-react"
            target="_blank"
          >
            babel-react
          </a>
        </td>
        <td
          align="Stars"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            156,788
          </span>
        </td>
        <td
          align="Forks"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            32,756
          </span>
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          align="Name"
          class=""
        >
          <a
            class="A-sc-70cat-0 ljzaKq"
            href="https://github.com/next"
            target="_blank"
          >
            next.js
          </a>
        </td>
        <td
          align="Stars"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            288
          </span>
        </td>
        <td
          align="Forks"
          class=""
        >
          <span
            class="sc-hKgILt gKBYtv"
          >
            34,998
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Manual Inputs > Trade Op main view component tests renders screen loader when the data is loading 1`] = `
<div
  class="sc-jSgupP kpIYLA"
>
  <div
    class="sc-eCssSg jyVNWN"
  />
</div>
`;
